types/gitremote.go

```diff
@@ -0,0 +1,156 @@
+package types
+
+import (
+	"fmt"
+	"strings"
+)
+
+type RemoteType int
+
+const (
+	SshRemote RemoteType = iota
+	HttpsRemote
+	HttpRemote
+	GitRawRemote
+	FileRemote
+)
+
+type GitRemote struct {
+	Url      string
+	Protocol string
+	Token    string
+	User     string
+	Password string
+	Domain   string
+	Path     string
+	Branch   string
+	Type     RemoteType
+}
+
+func NewGitRemote(url string) (r GitRemote) {
+	r.Url = url
+	r.Token, r.User, r.Password, r.Domain, r.Path, r.Protocol, r.Type = parseRemote(url)
+
+	return
+}
+
+func (r GitRemote) Validate() bool {
+	return r.Url != "" && r.Domain != "" && r.Path != ""
+}
+
+func (r GitRemote) String() (str string) {
+	if r.Validate() {
+		var template, auth string
+		n := 3
+
+		switch {
+		case r.Type == SshRemote || r.Type == GitRawRemote:
+			template = "%s%s:%s" // credentials@host:path
+			if r.User != "" {
+				auth = fmt.Sprintf("%s@", r.User)
+				if r.Password != "" {
+					auth = fmt.Sprintf("%s:%s@", r.User, r.Password)
+				}
+			}
+		case r.Type == FileRemote:
+			n = 2
+			template = "%s:///%s" // file:///path
+		case r.Type == HttpsRemote || r.Type == HttpRemote:
+			n = 4
+			template = "%s://%s%s/%s" // prot://credentials@domain/path
+			if r.Token != "" {
+				auth = fmt.Sprintf("x-access-token:%s@", r.Token)
+			} else if r.User != "" {
+				auth = fmt.Sprintf("%s:%s@", r.User, r.Password)
+			}
+		}
+		switch n {
+		case 2:
+			str = fmt.Sprintf(template, r.Protocol, r.Path)
+		case 3:
+			str = fmt.Sprintf(template, auth, r.Domain, r.Path)
+		case 4:
+			str = fmt.Sprintf(template, r.Protocol, auth, r.Domain, r.Path)
+		}
+	}
+	return
+}
+
+func parseRemote(s string) (token, user, password, domain, path, protocol string, t RemoteType) {
```

Comment:
Hmm I really think we should add a unit test for this, using this style:

https://github.com/golang/go/wiki/TableDrivenTests